### PR TITLE
Leachim/latent variable model

### PR DESCRIPTION
**Pull request by leachim**
Hi @cpfiffer, could you have a quick look and give me feedback on the general outline and ideas. There are still a few little rough edges, e.g. the math is not properly displayed and one of the outputs at the end is huge, but that's the general idea for the first tutorial.

Is the amount of text too little or too much? I wasn't sure about that. And I didn't talk much about model evaluation, maybe it's worth adding a bit on that, but it's a bit difficult when there are so many parameters...

### CODE REVIEW

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:316]: Okay, for this graph, you should note why it's different from the one above -- what did the batch effect do? To me, it looks like the classification problem got a lot harder. Make that explicit!

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:376]: Perfect, love it!

**Review by cpfiffer — COMMENTED** [review]
Wow, excellent stuff! Short and to the point, well-written code. I left some comments and suggestions as I went, but I know it's still a bit of a WIP and you can discard whatever notes I have if you wish. Let me know when you want another review!

**Comment by leachim** [comment]
I did a rebase on master, not sure this worked as it should. I am about to push the latest improvements, but I have got one question prior. Do you know why the math doesn't display in the output files? Do you need to have a certain package installed to support latex maths in the jmd files?

**Comment by cpfiffer** [comment]
Not sure what you mean exactly -- like the plots?

**Comment by leachim** [comment]
I was just not seeing math displayed correctly when I was viewing the md files in my browser. But I realize it's the same for the other notebooks, so might just be an issue with how the markdown files are displayed locally. So I think it's actually working already.

**Comment by leachim** [comment]
This should be pretty much ready now. I will ask someone else who hasn't seen it before to read through it, make some changes based on his feedback, and then maybe you can have another look before we release it more publicly.

**Comment by cpfiffer** [comment]
Perfect -- flag me whenever you want another review.

**Review by torfjelde — COMMENTED** [review]
This looks really good!
I had to look at your other PR, so figured I'd take a quick look at this too.

I've hadded some comments on the model-definitions that might be helpful. With those comments addressed it should also be possible to use Zygote, if one so desires.

And a more general question: maybe I've missed something, but is there a reason why we do Bayesian inference here rather than something simpler like MAP, i.e. is there a reason why we need the uncertainty estimate here? 
A couple of questions/thoughts regarding using displaying the uncertainty estimate:
- Is there any use in looking at the predictive posterior here? 
- Could we maybe plot a heatmap of the positions in the lower-dimensional space rather than the median?
- One possibility might be to compute a decision boundary in the lower-dimensional space for each MCMC sample and plot this, thus giving us a sense of the uncertainty of the separation in this space. Could also use animations, e.g. similar to what's done here https://turing.ml/dev/tutorials/03-bayesian-neural-network/#prediction-visualization.

The above might be particularly interesting when looking at the batch effects?

I'm not certain what sorts of timelines you have one these, so if the above suggestions are a bit too much, feel free to ignore me:)

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd
Comment (leachim) [reply tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:102]: no, just me being sloppy. I'll try to stick to the convention throughout.

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd
Comment (leachim) [reply tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:121]: I originally had more than 1 chain, and then I decided it was easier to just have one for the purpose of the tutorial. I'll put in some explanation. The problem is we are running a few chains (one per model), so I don't want any individual one to take too long.

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd
Comment (leachim) [reply tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:281]: :rofl: you caught me ;)

**Comment by leachim** [comment]
Thanks, that is all very useful stuff. I appreciate the suggestions, I'd definitely try to add some more visualization of the uncertainty. Regarding MAP estimates, I could add an example of how to obtain those, I think it's relatively straightforward in Turing. 
I'll write a bit more on this in the next few days.

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:6]: Could we have a little more motivation here?

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:49]: This stuff is a good motivation, and might be a good candidate to generalize and move to the top paragraph.

**Comment by leachim** [comment]
Okay, I hope this is ready for proper review now @cpfiffer

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:118]:
```suggestion
Here, we run the inference with the NUTS sampler. Feel free to try [different samplers](https://turing.ml/stable/docs/library/#samplers).
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:123]:
```suggestion
```

**Review by cpfiffer — APPROVED** [review]
Wow, spectacular work. I left two minor comments, but this is good to merge after the tests pass.

**Comment by leachim** [comment]
@cpfiffer okay if I merge this now?

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:38]:
```suggestion
using LinearAlgebra
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:41]: Also in this tutorial it seems a bit complicated to use two plotting frameworks?

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:108]: The constructor is deprecated.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:111]: Here as well.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (leachim) [reply tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:41]: Yes, I am only using statsplots to show one convergence plot of chains, everything else is done with VegaLite. I can remove that plot, but on the other hand, I don't think it's the end of the world having two different frameworks included.

**Comment by leachim** [comment]
@rikhuijzer I believe this is due to some bug with the caching (from what I saw in the slack channel). Please let me know when this is fixed and I am happy to update this and run it again.

**Comment by rikhuijzer** [comment]
> @rikhuijzer I believe this failing now is due to some bug with the caching (from what I saw in the slack channel). Please let me know when this is fixed and I am happy to update this and run it again.

I'm having a bit of trouble navigating the branch. Is this PR up-to-date with `master`? The fix from https://github.com/TuringLang/TuringTutorials/pull/278 sounds related. It should also fix the issue you had with CSV in the other PR.

Could you also avoid committing generated artifacts? You can do that by adding
```
**/11_probabilistic-pca
```
to the .gitignore and force removing the files manually.

**Comment by rikhuijzer** [comment]
And you are missing Compat entries from the Project.toml. Later, this repository will enable CompatHelper again. If you don't set the entries then CompatHelper will open one new PR for every package that you've included.

**Comment by rikhuijzer** [comment]
Sorry for the pedantic reviews, by the way. But as you might have noticed, the repository isn't as stable, easy to use and clean as it should be. So, for me at least, fixing existing problems has a higher priority than adding new tutorials.

**Comment by leachim** [comment]
Okay, I think I figured the compat thing out. I removed artifact, and it should be based off the latest master at the moment. Let's see if it runs through the checks now.

File: tutorials/11-probabilistic-pca/Project.toml
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/Project.toml:21]:
```suggestion
Bijectors = "0.9.8"
DataFrames = "1"
Distributions = "0.25"
KernelDensity = "0.6"
RDatasets = "0.7"
StatsPlots = "0.14"
Turing = "0.18"
VegaLite = "2.6"
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:109]: Are you using 2 spaces or 4 spaces for indentation? Here it is 2 and below it is 4. The Julia convention is 4 spaces. 2 would also be okay, but mixing them up doesn't help readability.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:158]: Thanks for adding the assertions :+1:

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:165]:
```suggestion
rename!(df_pca, Symbol.(["z"*string(i) for i in collect(1:n_genes)]))
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:218]:
```suggestion
# Extract parameter estimates for plotting - mean of posterior
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:223]:
```suggestion
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:236]:
```suggestion
rename!(df_pre, Symbol.(["z"*string(i) for i in collect(1:n_genes)]))
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:241]:
```suggestion
df_pre[!,:type] = repeat([1, 2], inner=n_cells÷2)
```
You're using no whitespace around equal symbols in keyword arguments pretty consistently.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:323]:
```suggestion
rename!(df_pre, Symbol.(["z"*string(i) for i in collect(1:d)]))
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:370]:
```suggestion
```

**Review by rikhuijzer — APPROVED** [review]
I don't know much about the statistical aspects of this tutorial, but if CI passes then I'm okay with this PR from the technical side.

EDIT: If you accept my suggestions, please use the batch apply feature from the "Files changed" tab.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd
Comment (leachim) [reply tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:109]: I have fixed it to 2 whitespaces throughout now.

**Comment by rikhuijzer** [comment]
In response to the failing CI. Did you see my comment above about merging `master` into this branch?

> I'm having a bit of trouble navigating the branch. Is this PR up-to-date with master? The fix from #278 sounds related.

**Comment by leachim** [comment]
> In response to the failing CI. Did you see my comment above about merging `master` into this branch?
> 
> > I'm having a bit of trouble navigating the branch. Is this PR up-to-date with master? The fix from #278 sounds related.

Yes, I saw this, but looking at https://github.com/TuringLang/TuringTutorials/network I think this branch is up-to-date with the master branch. I rebased it at least twice to keep it up to date with the latest master.

**Comment by rikhuijzer** [comment]
> > In response to the failing CI. Did you see my comment above about merging `master` into this branch?
> > > I'm having a bit of trouble navigating the branch. Is this PR up-to-date with master? The fix from #278 sounds related.
> 
> Yes, I saw this, but looking at https://github.com/TuringLang/TuringTutorials/network I think this branch is up-to-date with the master branch. I rebased it at least twice to keep it up to date with the latest master.

I'm sorry. You're right. I'm running the build now locally and precompilation succeeds here just fine. Do you know whether VegaLite needs some extra settings to run headless maybe?

**Comment by leachim** [comment]
> build("11-probabilistic-pca")

Okay, let me check that. I tested it on my machine before the latest commits and it was working all fine. I'll rerun the latest commit and see if there are any issues.

**Comment by leachim** [comment]
> > build("11-probabilistic-pca")
> 
> Okay, let me check that. I tested it on my machine before the latest commits and it was working all fine. I'll rerun the latest commit and see if there are any issues.

So this runs perfectly fine on my machine. I can build it all right. What kind of issue do you get when running it?

**Comment by rikhuijzer** [comment]
> > > build("11-probabilistic-pca")
> > 
> > 
> > Okay, let me check that. I tested it on my machine before the latest commits and it was working all fine. I'll rerun the latest commit and see if there are any issues.
> 
> So this runs perfectly fine on my machine. I can build it all right. What kind of issue do you get when running it?

For me, it failed on two different computers. The first error is "Failed to precompile Turing". But, now that I've run `build_folder("11-probabilistic-pca")` directly, my computer is already humming for 30 minutes, so I guess precompilation worked out.

I hope this information can help you, because I really have to work on some other things now.

**Comment by rikhuijzer** [comment]
After running a few more times locally and inspecting the logs, I think we are running into https://github.com/JuliaArrays/ArrayInterface.jl/issues/206.

**Comment by torfjelde** [comment]
> One possible fix would be to add TuringTutorials as a dependency to all the tutorials and call weave_file / build from there.

This seems like it would probably be a good idea. We'd just run `Pkg.develop("path-to-TuringTutorials"); using TuringTutorials: TuringTutorials` after each env activation, right?

**Comment by rikhuijzer** [comment]
> > One possible fix would be to add TuringTutorials as a dependency to all the tutorials and call weave_file / build from there.
> 
> This seems like it would probably be a good idea. We'd just run `Pkg.develop("path-to-TuringTutorials"); using TuringTutorials: TuringTutorials` after each env activation, right?
> 
> EDIT: We could move the `build.jl` and related to `scripts/` folder and have those use `TuringTutorials` rather than be part of it.

If you mean:
```
Pkg.develop("path-to-a-tutorial")
using TuringTutorials: build
build("path-to-a-tutorial")
```
then yes.

Sorry, but I don't understand your last sentence about moving the files.

**Comment by leachim** [comment]
let me know if you have fixed the issue with the build system, happy to merge the two tutorials in then.

**Pull request merged**
